authenticate.js

```diff
@@ -0,0 +1,42 @@
+function jsonParse(value, cb) {
+	var err = null, result
+	try {
+		result = JSON.parse(value)
+	} catch (e) {
+		err = e
+	}
+	cb(err, result)
+}
+
+module.exports = function a(emitter, tokenDb, tokenLock) {
+	return function authenticate(token, cb) { //cb(err, credentials)
+		if (!cb) cb = function () {}
+
+		if (!token) {
+			setTimeout(cb, 0, new Error('No token found'))
+		} else {
+			tokenLock.readLock(function (unlock) {

```

Comment:
We can use [this module](https://github.com/TehShrike/key-master) to map keys to mutexify instances.